File: interviews/python/rotationalCipher.py
def rotationalCipher(input, rotationFactor):
    alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
    digits = '0123456789'
    encrypted = ''
    for char in input:
        plainChar = char.upper()
        isUpperCase = plainChar == char
        if plainChar in alphabet:
            alphabetIndex = alphabet.index(plainChar)
            encryptedIndex = (alphabetIndex + rotationFactor) % len(alphabet)
            if isUpperCase:
                encrypted += alphabet[encryptedIndex]
            else:
                encrypted += alphabet[encryptedIndex].lower()
            continue
        if plainChar in digits:
            digitsIndex = digits.index(plainChar)
            encryptedIndex = (digitsIndex + rotationFactor) % len(digits)
            encrypted += digits[encryptedIndex]
            continue
        encrypted += char
    return encrypted

tests = [
    { 'input': 'All-convoYs-9-be:Alert1.', 'rotationFactor': 4, 'expected': 'Epp-gsrzsCw-3-fi:Epivx5.' },
    { 'input': 'abcdZXYzxy-999.@', 'rotationFactor': 200, 'expected': 'stuvRPQrpq-999.@' }
]

for test in tests:
    result = rotationalCipher(test['input'], test['rotationFactor'])
    print('input:', test['input'], 'rotationFactor:', test['rotationFactor'], 'expected:', test['expected'], 'result:', result)

File: interviews/javascript/rotationalCipher.js

function rotationalCipher(input, rotationFactor) {
    const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    const digits = '0123456789';
    let encrypted = '';
    // O(n) time
    for (let i = 0; i < input.length; i++) {
        const textChar = input[i].toUpperCase();
        const isUpperCase = textChar == input[i];
        // check for alphabetic character
        const alphabetIndex = alphabet.indexOf(textChar);
        if (alphabetIndex >= 0) {
            const encryptedIndex = (alphabetIndex + rotationFactor) % alphabet.length;
            const encryptedChar = alphabet[encryptedIndex];
            encrypted += isUpperCase ? encryptedChar : encryptedChar.toLowerCase();
            continue;
        }

        // check for a digit
        const digitsIndex = digits.indexOf(textChar);
        if (digitsIndex >= 0) {
            const encryptedIndex = (digitsIndex + rotationFactor) % digits.length;
            const encryptedChar = digits[encryptedIndex];
            encrypted += encryptedChar;
            continue;
        }

        // fallback, just append
        encrypted += textChar;
    }
    return encrypted;
}

const tests = [
    { input: 'All-convoYs-9-be:Alert1.', rotationFactor: 4, expected: 'Epp-gsrzsCw-3-fi:Epivx5.' },
    { input: 'abcdZXYzxy-999.@', rotationFactor: 200, expected: 'stuvRPQrpq-999.@' }
];

for (const test of tests) {
    const result = rotationalCipher(test.input, test.rotationFactor);
    console.log('input:', test.input, 'rotationFactor:', test.rotationFactor, 'expected:', test.expected, 'result:', result);
}